gt/lua_lpm.c

```diff
@@ -0,0 +1,375 @@
+
+static int
+l_str_to_prefix(lua_State *l)
+{
+	int ret;
+	struct ipaddr ip_addr;
+
+	/* First argument must be an IP prefix string. */
+	const char *prefix_str = luaL_checkstring(l, 1);
+
+	if (lua_gettop(l) != 1)
+		luaL_error(l, "Expected one argument, however it got %d arguments",
+			lua_gettop(l));
+
+	ret = parse_ip_prefix(prefix_str, &ip_addr);
+	if (ret < 0 || ip_addr.proto != ETHER_TYPE_IPv4)
+		luaL_error(l, "gk: failed to parse an IPv4 prefix");
+
+	lua_pushinteger(l, ip_addr.ip.v4.s_addr);
+	lua_pushinteger(l, ret);
+
+	return 2;
+}
+
+#define CTYPE_STRUCT_IN6_ADDR_REF "struct in6_addr &"
+
+static int
+l_str_to_prefix6(lua_State *l)
+{
+	int ret;
+	struct ipaddr ip_addr;
+	struct in6_addr *cdata;
+	uint32_t correct_ctypeid_in6_addr;
+
+	/* First argument must be an IP prefix string. */
+	const char *prefix_str = luaL_checkstring(l, 1);
+
+	if (lua_gettop(l) != 1)
+		luaL_error(l, "Expected one argument, however it got %d arguments",
+			lua_gettop(l));
+
+	ret = parse_ip_prefix(prefix_str, &ip_addr);
+	if (ret < 0 || ip_addr.proto != ETHER_TYPE_IPv6)
+		luaL_error(l, "gk: failed to parse an IPv6 prefix");
+
+	correct_ctypeid_in6_addr = luaL_get_ctypeid(l,
+		CTYPE_STRUCT_IN6_ADDR_REF);
+	cdata = luaL_pushcdata(l, correct_ctypeid_in6_addr,
+		sizeof(struct in6_addr));
+	*cdata = ip_addr.ip.v6;
+
+	lua_pushinteger(l, ret);
+
+	return 2;
+}
+
+#define LUA_LPM_TNAME "gt_lpm"
+
+static int
+l_new_lpm(lua_State *l)
+{
+	struct rte_lpm_config lpm_conf;
+	struct rte_lpm **p_lpm;
+	static rte_atomic32_t identifier = RTE_ATOMIC32_INIT(0);
+
+	memset(&lpm_conf, 0, sizeof(lpm_conf));
+
+	/* First argument must be a Lua number. */
+	lpm_conf.max_rules = luaL_checknumber(l, 1);
+
+	/* Second argument must be a Lua number. */
+	lpm_conf.number_tbl8s = luaL_checknumber(l, 2);
+
+	if (lua_gettop(l) != 2)
+		luaL_error(l, "Expected two arguments, however it got %d arguments",
+			lua_gettop(l));
+
+	p_lpm = lua_newuserdata(l, sizeof(struct rte_lpm *));
+	*p_lpm = init_ipv4_lpm("gt_", &lpm_conf, rte_socket_id(),
+		rte_lcore_id(), rte_atomic32_add_return(&identifier, 1));
+	if (unlikely(*p_lpm == NULL))
+		luaL_error(l, "gt: failed to initialize the IPv4 LPM table for Lua policies");
+
+	luaL_getmetatable(l, LUA_LPM_TNAME);
+	lua_setmetatable(l, -2);
+
+	return 1;
+}
+
+static int
+l_lpm_add(lua_State *l)
+{
+	int ret;
+
+	/* First argument must be of type struct rte_lpm **. */
+	struct rte_lpm *lpm =
+		*(struct rte_lpm **)luaL_checkudata(l, 1, LUA_LPM_TNAME);
+
+	/*
+	 * Second argument must be a Lua number.
+	 * @ip must be in network order.
+	 */
+	uint32_t ip = luaL_checknumber(l, 2);
+
+	/* Third argument must be a Lua number. */
+	uint8_t depth = luaL_checknumber(l, 3);
+
+	/* Fourth argument must be a Lua number. */
+	uint32_t label = luaL_checknumber(l, 4);
+
+	if (lua_gettop(l) != 4)
+		luaL_error(l, "Expected four arguments, however it got %d arguments",
+			lua_gettop(l));
+
+	ret = rte_lpm_add(lpm, ip, depth, label);
```

Comment:
Shouldn't we replace `ip` with `ntohl(ip)`?

Reply:
Good catch!